pytorch_lightning/trainer/supporters.py

```diff
@@ -119,27 +122,249 @@ def mean(self):
+        - all_gather_predictions: Gather predictions accross multiple processses.
+
+    """
+
+    ID_KEY = 'id'
+
+    def __init__(self, global_rank: int, world_size: int, all_gather_fn: Callable):
         self.global_rank = global_rank
         self.world_size = world_size
-        self.predictions = {}
-        self.num_predictions = 0
+        self.all_gather_fn = all_gather_fn
+        self._legacy_predictions = {}
+        self._predictions = {stage: {} for stage in LoggerStages}
+
+    @property
+    def predictions(self) -> Dict:
+        return self._predictions[self.current_stage]
+
+    def _cache_prediction(
+        self,
+        predictions: List[Dict],
+        dl_idx: int,
+        batch_indices: List[int],
+        enable_predict_auto_id: bool
+    ) -> None:
+
+        cache = self.predictions
+
+        cache.setdefault(dl_idx, {})
+
+        def convert(value):
+            return value.cpu().tolist()
+
+        for batch_idx, pred in enumerate(predictions):
+            if enable_predict_auto_id:
+                sample_id = batch_indices[batch_idx]
+
+            else:
+                if self.ID_KEY in pred:
+                    sample_id = pred[self.ID_KEY]
+                    if not isinstance(sample_id, (int, float)):
+                        raise MisconfigurationException(
+                            f"`id` key should be either a int or float. Found {type(sample_id)}."
+                        )
+                else:
+                    raise MisconfigurationException(
+                        f"The predictions dict requires an `{self.ID_KEY}` key."
+                    )
+
+            if sample_id in cache[dl_idx]:
+                raise MisconfigurationException(
+                    "Prediction Collection doesn't support multiple predictions for one sample yet.")
+
+            if enable_predict_auto_id:
+                cache[dl_idx][sample_id] = {
+                    self.ID_KEY: sample_id, "predictions": pred}
+            else:
+                cache[dl_idx][sample_id] = pred
+
+            # apply convert to store memory
+            cache[dl_idx][sample_id] = apply_to_collection(cache[dl_idx][sample_id], torch.Tensor, convert)
+
+    def cache(
+        self,
+        predictions: List,
+        dl_idx: int,
+        batch_indices: List[int],
+        current_stage: str,
+        enable_predict_auto_id: bool
+    ) -> None:
+        """
+        This function expects predictions to be a list of tensors or dictionary of tensors.
+        Example::
+
+            self.add_predictions(predictions)
+        """
+        if predictions is None or (isinstance(predictions, list) and not predictions):
+            return
+
+        self.current_stage = current_stage
+
+        if enable_predict_auto_id:
+            assert isinstance(predictions, (list, torch.Tensor))
+            if batch_indices is None:
+                return
+            assert len(predictions) == len(batch_indices)
```

Comment:
```suggestion
            assert len(predictions) <= len(batch_indices)
```
but as it is a public API, convert it to raise an exception

Reply:
Done !